doc/appendices/development.rst

```diff
@@ -8,4 +8,400 @@
+offers good suggestions.
+
+.. _ISO C++ guidelines: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines
+.. _Google style guide: https://google.github.io/styleguide/cppguide.html
+
+
+Test thoroughly
+---------------
+
+Functions should use programmatic assertions whenever assumptions are made:
+
+- Use the ``CELER_EXPECT(x)`` assertion macro to test preconditions about
+  incoming data or initial internal states
+- Use ``CELER_ASSERT(x)`` to express an assumption internal to a function (e.g.
+  "this index is not out of range of the array")
+- Use ``CELER_ENSURE(x)`` to mark expectations about data being returned from a
+  function and side effects resulting from the function.
+
+Additionally, user-provided data and potentially volatile runtime conditions
+(such as the presence of an environment variable) should be checked with
+the ``CELER_VALIDATE(x, << "streamable message")`` macro. See :ref:`corecel`
+for more details about these macros.
+
+Each class must be thoroughly tested with an independent unit test in the
+`test` directory.  For complete coverage, each function of the class must have
+at least as many tests as the number of possible code flow paths (cyclomatic
+complexity).
+
+Implementation detail classes (in the ``celeritas::detail`` namespace, in
+``detail/`` subdirectories) are exempt from the testing requirement, but
+testing the detail classes is a good way to simplify edge case testing compared
+to testing the higher-level code.
+
+Style guidelines
+================
+
+Having a consistent code style makes it more readable and maintainable. (For
+example, you don't have to guess whether a symbol is a function or class.)
+
+As a historical note, many of the style conventions in Celeritas derive from
+the `Draco project style`_ of which `Tom Evans`_ was primary author and which
+became the style standard for the GPU-enabled Monte Carlo code `Shift`_.
+
+.. _Draco project style: https://github.com/lanl/Draco/wiki/Style-Guide
+.. _Tom Evans: https://github.com/tmdelellis
+.. _Shift: https://doi.org/10.1016/j.anucene.2019.01.012
+
+Formatting
+----------
+
+Formatting is determined by the clang-format file inside the top-level
+directory. One key restriction is the 80-column limit, which enables multiple
+code windows to be open side-by-side. Generally, statements longer than 80
+columns should be broken into sub-expressions for improved readability anyway
+-- the ``auto`` keyword can help a lot with this.
+
+There's a certain amount of decorations (separators, Doxygen comment structure,
+etc.) that is standard throughout the code. Use the ``celeritas-gen.py`` script
+(in the ``scripts/dev`` directory) to generate skeletons for new files, and use
+existing source code as a guide to how to structure the decorations.
+
+Symbol names
+------------
+
+Functions should be verbs; classes should be names. As in standard Python
+(PEP-8-compliant) code, classes should use ``CapWordsStyle`` and variables use
+``snake_case_style``.
```

Comment:
You may want to add a mention of the convention of members of structs and classes.